truffle/com.oracle.truffle.api/src/com/oracle/truffle/api/source/SourceSection.java

```diff
@@ -76,6 +83,30 @@
         this.startColumn = startColumn;
         this.charIndex = charIndex;
         this.charLength = charLength;
+        this.tags = tags;
+        assert tagsAreNonNullAndInterned(tags) : "All tags set for a source section must be interned and non-null.";
+    }
+
+    @SuppressFBWarnings("ES_COMPARING_STRINGS_WITH_EQ")
+    private static boolean tagsAreNonNullAndInterned(String[] tags) {
+        for (int i = 0; i < tags.length; i++) {
+            if (tags[i] == null) {
+                return false;
+            }
+            if (tags[i].intern() != tags[i]) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    /**
+     * Returns a set of tags that are associated with this source section as array. The returned
+     * array must be considered read-only. All elements of the array were asserted to be non-null
+     * and interned for faster comparison.
+     */
+    public String[] getTags() {
```

Comment:
I would suggest to hide this getter behind Accessor. There is no real usecase for our languages or instruments to call this method and no example does it.

If you leave the method public, it will be harder to get rid of the "tags as strings" in the future.